Clean up error checking in params (#210)
lanl/parthenon: docs/README.md

Macros for causing execution to throw an exception are provided [here](../src/utils/error_checking.hpp)
* PARTHENON_REQUIRE(condition, message) exits if the condition does not evaluate to true.
* PARTHENON_REQUIRE_THROWS(condition, message) throws a `std::runtime_error` exception if the condition does not evaluate to true.
* PARTHENON_FAIL(message) always exits.
* PARTHENON_THROW(message) throws a runtime error.
* PARTHENON_DEBUG_REQUIRE(condition, message) exits if the condition does not evaluate to true when in debug mode.
* PARTHENON_DEBUG_REQUIRE_THROWS(condition, message) throws if the condition does not evaluate to true when in debug mode.
* PARTHENON_DEBUG_FAIL(message) always exits when in debug mode.
* PARTHENON_DEBUG_THROW(message) throws a runtime error when in debug mode.

All macros print the message, and filename and line number where the
macro is called. PARTHENON_REQUIRE also prints the condition. The
macros take a `std::string`, a `std::stringstream`, or a C-style
string. As a rule of thumb:

> - Use the exception throwing versions in non-GPU, non-performance critical code.

Reviewer: ```suggestion
- Use the exception throwing versions outside of kernels/parallel regions and non-performance critical code.
```

> - On GPUs and in performance-critical sections, use the non-throwing

Reviewer: ```suggestion
- Within parallel regions and in performance-critical sections, use the non-throwing
```
